First prototype of the dune-precice module (dune-adapter) for preCICE

maxfirmbach opened the pull request:
Hello preCICE-team and developers,

this is a prototype of a module inside the DUNE framework, representing an adapter for the preCICE-library.
For now this module is mainly used for FE calculations, enabling multi-physics simulations within DUNE.

Literature references can be found in the README inside the dune-precice folder.

maxfirmbach committed:
inital push to develop

maxfirmbach committed:
update README

maxfirmbach committed:
update README

uekerman commented:
Thanks for the contribution, @maxfirmbach :tada: 
Why did you put everything in the subfolder `dune-precice`? Feel free to completely overwrite the preliminary README that is already on the main branch.

maxfirmbach committed:
move description to outer Readme

MakisH commented:
@maxfirmbach thank you very much for the contribution! Before reviewing it, do you maybe have a case that we can run it with? If you want, you can also open a separate PR in the tutorials.

ajaust commented:
This looks like a great addition to the adapter repositories. I have worked on an adapter for [DuMuX](https://dumux.org/) which shares quite some functionality with the DUNE adapter. I tried to give an extensive overview in #2 .

maxfirmbach commented:
@MakisH I can create a tutorial case and will do a PR in the tutorials (might however take some time). To run a test case the DUNE core modules and some extension modules are needed and the first installation of the whole DUNE environment can be a bit cumbersome.

ajaust commented:
Regarding the installation of DUNE. There is a [Spack recipe for DUNE](https://gitlab.dune-project.org/spack/dune-spack). Alternatively, there is an [installation script for DuMuX](https://git.iws.uni-stuttgart.de/dumux-repositories/dumux/blob/master/bin/installdumux.py). The installation script also installs DUNE dependencies. The dependency list in line 134 needs to be adapted for the DUNE adapter to contain `common`, `functions` and `istl` (`functions` is currently missing as DuMuX does not need it). If one is not interested in DuMuX, one can comment out the corresponding parts or simply remove the `dumux` directory after the installation.

maxfirmbach committed:
Add perpendicular-flap tutorial case

maxfirmbach commented:
I added a case, where Dune acts as solid-participant in the perpendicular-flap tutorial case. The corresponding folder inside the tutorial directory should also be ready.

davidscn commented on README.md:
```diff
@@ -1,2 +1,28 @@
+https://github.com/maxfirmbach/dune-elastodynamics
+
+The adapter can be build with:
+`<path-to-dune-common/bin/dunecontrol> --current all`
```
I installed dune from apt, where `dunecontrol` is in the global searchpath:
```suggestion
`dunecontrol --current all`
```

davidscn replied:
Maybe link here also to the module installation instructions provided by DUNE itself. Is the `dunecontrol make install` command missing?

maxfirmbach replied:
I guess that are the instructions on the website: https://www.dune-project.org/doc/installation/

davidscn commented on README.md:
```diff
@@ -1,2 +1,28 @@
+on dune-elastodynamics providing the necessary tools for structural simulation.
+
+The dune-elastodynamics module can be found here:
+https://github.com/maxfirmbach/dune-elastodynamics
```
Cloning another repository is somewhat tedious. Would it be possible to include the files here? Or add a `git submodule`?

davidscn replied:
The `foamgrid` module is a dependency by the elastodynamics module and non-default module (at least not available via apt). This should be documented somewhere.

davidscn replied:
Building of the elastodynamics module fails with the first error message being:

```bash
In file included from dune-adapter/dune-elastodynamics/dune/elastodynamics/timesteppers/newmark.hh:10,
                 from dune-adapter/dune-elastodynamics/src/Example03.cc:36:
dune-adapter/dune-elastodynamics/dune/elastodynamics/preconditioners/distributedscalarproduct.hh: In instantiation of ‘class DistributedScalarProduct<Dune::GridView<Dune::UGGridLeafGridViewTraits<const Dune::UGGrid<2> > >, Dune::BlockVector<Dune::FieldVector<double, 2> > >’:
dune-adapter/dune-elastodynamics/dune/elastodynamics/timesteppers/newmark.hh:55:24:   required from ‘void Dune::Newmark<GridView, MatrixType, VectorType>::initialize(VectorType&, VectorType) [with GridView = Dune::GridView<Dune::UGGridLeafGridViewTraits<const Dune::UGGrid<2> > >; MatrixType = Dune::BCRSMatrix<Dune::FieldMatrix<double, 2, 2> >; VectorType = Dune::BlockVector<Dune::FieldVector<double, 2> >]’
dune-adapter/dune-elastodynamics/src/Example03.cc:110:46:   required from here
dune-adapter/dune-elastodynamics/dune/elastodynamics/preconditioners/distributedscalarproduct.hh:32:22: error: ‘typename Dune::ScalarProduct<Vector>::field_type DistributedScalarProduct<GridView, Vector>::dot(const Vector&, const Vector&) const [with GridView = Dune::GridView<Dune::UGGridLeafGridViewTraits<const Dune::UGGrid<2> > >; Vector = Dune::BlockVector<Dune::FieldVector<double, 2> >; typename Dune::ScalarProduct<Vector>::field_type = double]’ marked ‘override’, but does not override
```

maxfirmbach replied:
So I have one folder, where all the dune-modules are in (I cloned all of them).
I have set the `DUNE_CONTROL_PATH` to this folder inside my bashrc e.g. `export DUNE_CONTROL_PATH="/home/a11bmafi/Library/dune"`. I never installed the binary packages. To build all the modules I go inside my dune folder and give the path to dunecontrol e.g. `dune-common/bin/dunecontrol all`. A single module can be build by changing into the dune-module folder and do something like `../dune-common/bin dunecontrol --current all` at least that's how I'm doing it .... maybe there's also a better way.

davidscn replied:
Do you have a requirement on a specific version? Running `dunecontrol` works in the first place, but it is a compiler error as you can see.

maxfirmbach replied:
I use version 2.7 ... Two students also installed it and for them everything worked fine.

davidscn replied:
Hm I installed version `2.6`. Do I need `dune-python`?

maxfirmbach replied:
I haven't installed `dune-python` ... I guess it wont work with version `2.6`, sorry, I also need to add to use at least `2.7`.
For Ubuntu, there should also be a `2.7` version as binary package.

ajaust replied:
You can add the required DUNE module versions to the `dune.module` file. [DuMuX is doing it that way as well.](https://git.iws.uni-stuttgart.de/dumux-repositories/dumux/-/blob/master/dune.module)

davidscn replied:
@ajaust or @maxfirmbach is it possible to clone/download all core modules at once or do I really need to clone all of them manually?

davidscn replied:
>  For Ubuntu, there should also be a 2.7 version as binary package.

My ubuntu version is not recent enough, unfortunately.

maxfirmbach replied:
I cloned them one by one to be honest.

ajaust replied:
I do recommend the [`installdumux.py`](https://git.iws.uni-stuttgart.de/dumux-repositories/dumux/blob/master/bin/installdumux.py) script. I use it for my installation as it is easy to extend for the "standard" DUNE packages. You mainly need to adapt this line https://git.iws.uni-stuttgart.de/dumux-repositories/dumux/blob/master/bin/installdumux.py#L134 and remove some of the dumux related stuff.

You could try my script [installdune-2.7.py.txt](https://github.com/precice/dune-adapter/files/7381827/installdune-2.7.py.txt) that is based on the DuMuX script. I added the `.txt` file ending so I could upload it to GitHub. You still need to change line 90 for the modules that you actually want/need and maybe remove the part with the `cmake.opts` file if you do not want it.

davidscn replied:
```bash
ERROR: version mismatch.
       dune-precice requires 'dune-elastodynamics >= 2.7',
       but only 'dune-elastodynamics' = '1.0' is available.
```

maxfirmbach replied:
I adjusted the version in the dune.module file. You can also just manually delete the (2.7) after dune-elastodynamics in the dune-precice dune.module file.

davidscn reviewed (changes requested):
I tried to install the module but there was a problem (see below)

The directory structure is actually fine (`dune-precice`).

maxfirmbach committed:
Add dune installation to readme

maxfirmbach committed:
set required dune version to 2.7

maxfirmbach committed:
Adjust dune-elastodynamics version

davidscn commented on dune-precice/src/dune-perpendicular-flap.cc:
```diff
@@ -0,0 +1,222 @@
+
+  // setup looping
+  double t = 0.0;
+  double dt = 0.0001;
```
Why do you need to select such a small time-step size? The tutorial uses actually `1e-2`?!

maxfirmbach replied:
An explicit RKN-method is used in this case and for larger time steps I had difficulties with stability. I guess the tutorial cases use an implicit method.

davidscn replied:
Ah ok. This probably explains the deviation of the displacement amplitude as well (I can run a brief comparison tomorrow). You don't have any other time-stepping methods available, right? All in all, I can run the tutorial case andI will have a closer look at the code tomorrow.

maxfirmbach replied:
There should also be a family of Newmark methods (might have to clean that up a bit) and adaptive explicit RKN methods. But great that the thing runs at least.

maxfirmbach committed:
added implicit timestepping case

maxfirmbach commented:
I added a implicit time stepping case `dune-perpendicular-flap-implicit.cc` using Newmark's method and a time step size of 0.01, which should be the same as the openfoam fluid case.

I also cleaned up the time stepping method in dune-elastodynamics so before trying the new case, get a fresh pull of that please. I did a simulation, which looked fine ... I hope it works for you too.

davidscn commented:
> I added a implicit time stepping case dune-perpendicular-flap-implicit.cc using Newmark's method and a time step size of 0.01, which should be the same as the openfoam fluid case.
>
> I also cleaned up the time stepping method in dune-elastodynamics so before trying the new case, get a fresh pull of that please. I did a simulation, which looked fine ... I hope it works for you too.

Yes, this works as well and it is faster and more compliant to the other tutorials. However, there is still some deviation between our other tutorials and the dune code here, even with the implicit time-stepper (as you probably know)

![](https://user-images.githubusercontent.com/33414590/138240160-36c6b3db-c3cf-48f0-a128-dbe24ff6604e.png)

Do you have any ideas about the reasoning? Method-wise, you use continuous finite-elements of polynomial degree 2 and an implicit Newmark time-stepping with `gamma=0.5` and `beta=0.25`, right?

maxfirmbach commented:
Yes, I use continuous finite-elements of polynomial degree 2 and an implicit Newmark time-stepping with gamma=0.5 and beta=0.25. To be honest, I'm not sure why the result is different ... I already had a close look at all parts and to my best knowledge the structural part is correct. Still I couldn't figure out till now where the deviation comes from ...

Maybe a short remark, there's a file called `hooketensor.hh`. Due to the fact that the simulation is 2D there can be a plane stress or plane strain, right now plane stress is active ... maybe that also makes a difference.

davidscn commented:
Ah that's nice, we use `plane-strain`. Switching this for the dune adapter give actually the correct results:

![](https://user-images.githubusercontent.com/33414590/138264373-6eb2dfe2-a81c-4fb7-b273-ff694d2528f7.png)

so please change this in your elastodynamics module. I will wait for the feedback of @ajaust here for now.

maxfirmbach commented:
That looks really nice! I can't just change it ... cause for beam bending problems I need plane stress ... but I will find a way to call the right one.

Thanks for the help so far!

ajaust commented on README.md:
```diff
@@ -1,2 +1,31 @@
 # dune-adapter
```
I think the whole repository should be called `dune-precice` to follow standard DUNE notation. At the same time the structure of the repository should be adapted accordingly. Namely, all the content in the folder `dune-precice` should be in the root of the repository.

ajaust commented on dune-precice/dune.module:
```diff
@@ -0,0 +1,10 @@
+Version: 1.0
+Maintainer: max.firmbach@unibw.de
+#depending on
+Depends: dune-common (>= 2.7) dune-istl (>= 2.7) dune-functions (>= 2.7) dune-elastodynamics
```
Currently, this repository depends on `dune-elastodynamics` which does not make sense to me. This increases the number of dependencies while I would vote for `dune-precice` being a baseline implementation that can be used in other DUNE modules with minimum dependencies. Why would I care installing `dune-elastodynamics` if I do want to use `dune-precice` for a completely different type of application?

Therefore I would suggest
```suggestion
Depends: dune-common (>= 2.7) dune-istl (>= 2.7) dune-functions (>= 2.7)
```
and adding a simple example (dummysolver?) that purely relies on `dune-precice` and its dependencies.

I can see three scenarios at the moment:

1. Relaxing the current dependency  Make the current example a suggested dependency and prevent automatic building of the example.
  ```suggestion
  Depends: dune-common (>= 2.7) dune-istl (>= 2.7) dune-functions (>= 2.7)
  Suggests: dune-elastodynamics (>=1.0)
  ```
In this case I would also rename the directory. `src/` is not very verbose. Something like `examples/elastodynamic/` would be better.

2. Move all examples that require further dependencis into its one DUNE module (`dune-precice-examples`?).
3. The current example could live in `dune-elastodynamics` and one could add `dune-precice` as a suggested dependency there.

maxfirmbach replied:
The `dune-precice` module is now only depending on the core modules. I moved the examples to `dune-precice-howto` as this naming convention is also used for tutorial modules of certain core modules.

maxfirmbach replied:
Considering `dune-precice-examples` I was thinking about cutting the dependency to dune-elastodynamics completely in the future and replacing it by dune-pdelab. A linear-elasticity example should be rather easy to construct there.

ajaust commented on dune-precice/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+add_subdirectory(cmake/modules)
+
+# find precice library
+find_package(precice REQUIRED CONFIG)
```
Does the adapter work with any preCICE version (v1 vs. v2)? If not consider adding a minimum version.
```suggestion
find_package(precice 2.0.0 REQUIRED CONFIG)
```

maxfirmbach replied:
I added this to the dune-precice cmake, as we only want to work with version 2.0.

ajaust commented on dune-precice/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+dune_enable_all_packages()
+
+add_subdirectory(src)
+add_subdirectory(dune)
```
Why not move the adapter in a directory called `dune-precice`? Than the use DUNE module is more verbose in the files that include it.

maxfirmbach replied:
That's a good idea, i think something similar is done in DuMux too.

ajaust commented on dune-precice/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+
+dune_enable_all_packages()
+
+add_subdirectory(src)
```
As mentioned above, I do not think that `src/` is a good name for examples.

maxfirmbach replied:
The tutorial cases are now placed in a separated DUNE module inside `examples/`.

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      
+      std::vector<VectorType> old_state_data;
+      double old_time;
+      int old_iter;
```
Why do these member variables not end with an underscore `_`? The member `mpiRank_` does though?!
```suggestion
      const std::string mesh_name_;
      const std::string read_data_name_;
      const std::string write_data_name_;
    
      int mesh_id_;
      int read_data_id_;
      int write_data_id_;
      
      int n_interface_nodes_;
      std::vector<double> interface_nodes_positions_;
      std::vector<int>    interface_nodes_ids_;
      
      std::vector<double> read_data_;
      std::vector<double> write_data_;
      
      std::vector<VectorType> old_state_data_;
      double old_time_;
      int old_iter_;
```

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      
+    public:
+          
+      double precice_timestep_length;
```
Why is this a public member?

maxfirmbach replied:
You're right, this is not an optimal way to do it ;). I'll return the value directly and work with an additional variable `precice_dt` in the example itself.

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      void advance(double computed_timestep_length);
+       
+      void save_current_state(std::vector<VectorType*> state_quantaties, double &current_time_value, int &current_iter_value);
+      void reload_old_state(std::vector<VectorType*> state_quantaties, double &current_time_value, int &current_iter_value);
```
- There is no need to pass trivial types such as `double` or `int` as reference. See [C++ core guidelines on this](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const)
- Why is the vector passed by values and not by (const) reference here?

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      int old_iter;
+      
+      void format_dune_to_precice(VectorType &dune_to_precice);
+      void format_precice_to_dune(VectorType &precice_to_dune);
```
- `format_dune_to_precice` and opposite function `format_precice_to_dune` should take const reference as argument to emphasize that this is not meant to change the passed vector.
- Functions name is not optimal since this does not just format, but also copy data around which is rather implicit. 
- Maybe it should be a free (in a `internal` or similar namespace) or static function. Then one could have two Vectors passed in. One as const reference (goes in does not change) and one as reference (goes in gets changed) to make the intent/usage of the finction more clear.
- The argument names are not clear. From my understanding `dune_to_precice` does not map anything, but it is the vector in DUNE format or preCICE format. Not sure.

maxfirmbach replied:
I changed this. I think the naming is much clearer now, but maybe there's an even better way to do it.

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      void initialize(const Basis& basis);
+       
+      void send_blockvector_data(VectorType &dune_to_precice);
+      void read_blockvector_data(VectorType &precice_to_dune);
```
As above, I do not find `dune_to_precice` and `precice_to_dune` very verbose.

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+          
+      double precice_timestep_length;
+    
+      CouplingInterface(const ParameterClass &parameters, BlockVector<FieldVector<bool, dim>>& boundaryDofs, int mpi_rank, int mpi_size);
```
- Why is `boundaryDofs` not passed as const reference?
```suggestion
      CouplingInterface(const ParameterClass &parameters, const BlockVector<FieldVector<bool, dim>>& boundary_dofs, int mpi_rank, int mpi_size);
```
- Why are `mpi_size` and `mpi_rank` not immutable? See [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#con1-by-default-make-objects-immutable)
- The line seems to be very long.

ajaust commented on dune-precice/src/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+find_package(precice REQUIRED CONFIG)
```
I did not check this file.

ajaust commented on dune-precice/src/dune-perpendicular-flap-implicit.cc:
```diff
@@ -0,0 +1,222 @@
+// -*- tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*-
```
I did not check this file.

ajaust commented on dune-precice/src/dune-perpendicular-flap.cc:
```diff
@@ -0,0 +1,222 @@
+// -*- tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*-
```
I did not check this file.

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+                                                                        BlockVector<FieldVector<bool,dim>>& boundaryDofs,
+                                                                        int mpi_rank, int mpi_size)
+    : precice(parameters.participant_name, parameters.config_file, mpi_rank, mpi_size)
+    , isBoundary(boundaryDofs)
```
- The variable `isBoundary` is initialized by a variable called `boundaryDofs` which, to me, has as slightly different meaning. Could we homogenize the naming of variables to make their meaning more clear?

maxfirmbach replied:
You're right, they are different things, I will change the naming.

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      void reload_old_state(std::vector<VectorType*> state_quantaties, double &current_time_value, int &current_iter_value);
+                                         
+      bool is_save_required();
+      bool is_load_required();
```
```suggestion
      bool is_save_required() const;
      bool is_load_required() const;
```
Mark functions that do not change any members as const.

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      void mark_load_fullfilled(); 
+        
+      bool is_coupling_ongoing();
+      bool is_time_window_complete();
```
```suggestion
      bool is_coupling_ongoing() const;
      bool is_time_window_complete() const;
```

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+    
+      precice::SolverInterface precice;
+      
+      BlockVector<FieldVector<bool, dim>> isBoundary;
```
```suggestion
      const BlockVector<FieldVector<bool, dim>> is_boundary;
```
- Why is it not const? It is initialized in the constructor and never written afterwards.
- Why is it not snake case?
- Is this a good name? What kind of boundaries does it mark? Domain boundary? The coupling boundaries? Something else?
```suggestion
      BlockVector<FieldVector<bool, dim>> is_coupling_boundary;
```

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+        
+    FieldVector<bool, dim> indicator;
+    indicator = true;
+    n_interface_nodes = std::count(isBoundary.begin(), isBoundary.end(), indicator);
```
- I like the use of the standard library! :smile: 
- Could we construct the temporary in place to make it shorter?
```suggestion
    n_interface_nodes = std::count(isBoundary.begin(), isBoundary.end(), FieldVector<bool, dim>{true});
```
Alternatively I would try to at least limit the scope of the temporary by using braces
```suggestion
    {
        FieldVector<bool, dim> indicator;
        indicator = true;
        n_interface_nodes = std::count(isBoundary.begin(), isBoundary.end(), indicator);
    }
```

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+    
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  template <typename Basis>
```
I find `Basis` vague. What kind of basis is meant? The base class? The polynomial basis? I assume the polynomial basis. Thus I would suggest
```suggestion
  template <typename PolynomialBasis>
```

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+	write_data_id = precice.getDataID(write_data_name, mesh_id);
+            
+    // interpolate boundary DOFS to coordinates
+    VectorType coordinatesBoundary;
```
- This variable does not follow the name style of others (camel case) vs snake case.
- Maybe a slightly different name would be more specifig.
```suggestion
    VectorType coordinates_of_boundary;
```

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      void format_dune_to_precice(VectorType &dune_to_precice);
+      void format_precice_to_dune(VectorType &precice_to_dune);
+      
+      int mpiRank_;
```
Why is this not snake case?
```suggestion
      int mpi_rank_;
```

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      CouplingInterface(const ParameterClass &parameters, BlockVector<FieldVector<bool, dim>>& boundaryDofs, int mpi_rank, int mpi_size);
+
+      template <typename Basis>
+      void initialize(const Basis& basis);
```
```suggestion
      template <typename PolynomialBasis>
      void initialize(const PolynomialBasis& polynomial_basis);
```

ajaust commented on README.md:
```diff
@@ -1,2 +1,31 @@
+For more detailed installation instructions have a look at:
+https://www.dune-project.org/doc/installation/
+
+## tutorial
```
```suggestion
## Tutorial
```

ajaust commented on README.md:
```diff
@@ -1,2 +1,31 @@
+A adapter module enabeling multiphysics simulation inside DUNE with the help of preCICE [1][2].
+
+The dune-adapter module depends on the DUNE core modules and
+on dune-elastodynamics, dune-foamgrid providing the necessary tools for structural simulation.
```
The adapter also depends on preCICE. This should be stated here. We should also state which versions of the DUNE modules and of preCICE are needed.

maxfirmbach replied:
I changed this part, still have to add the versions though.

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+// -*- tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*-
```
I like the name `CouplingInterface` much more than the `SolverInterface` naming of preCICE. :smile:

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+#include <dune/istl/io.hh>
+#include <dune/istl/bvector.hh>
+
+#include <dune/functions/functionspacebases/interpolate.hh>
```
- Should we order thins in a somewhat alphabetically order?
- I like the general order (STL, DUNE, precice)

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+ 
+    interpolate(basis, coordinatesBoundary, coordinate, isBoundary);
+
+    // set up precice interface
```
Does this comment add much information? It is directly followed by some commands that do not any calls to preCICE.

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+	read_data_id  = precice.getDataID(read_data_name, mesh_id);
+	write_data_id = precice.getDataID(write_data_name, mesh_id);
+            
+    // interpolate boundary DOFS to coordinates
```
Does this comment explain much?

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      template <typename Basis>
+      void initialize(const Basis& basis);
+       
+      void send_blockvector_data(VectorType &dune_to_precice);
```
```suggestion
      void send_blockvector_data(const VectorType &dune_to_precice);
```
Why is it not a const reference? Is there a risk the data I pass is being changed?

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+  template <int dim, typename VectorType, typename ParameterClass>
+  void CouplingInterface<dim, VectorType, ParameterClass>::read_blockvector_data(VectorType &precice_to_dune) {
+  
+    // modify something here for nonoverlapping meshes ?
```
Remove comment and create and issue instead?

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+  template <int dim, typename VectorType, typename ParameterClass>
+  void CouplingInterface<dim, VectorType, ParameterClass>::send_blockvector_data(VectorType &dune_to_precice) {
+
+    // modify something here for nonoverlapping meshes ?
```
Remove comment and create and issue instead?

ajaust commented on dune-precice/dune/precice/couplinginterface.hh:
```diff
@@ -0,0 +1,257 @@
+      for(int j=0; j<dim; j++) {
+        if(isBoundary[i][j]) { write_data[iteration_count] = dune_to_precice[i][j]; ++iteration_count; }
+      }
+    }  
```
- Could we avoid raw loops here and instead used `std::for_each`, for example? See [Slide 3](https://sean-parent.stlab.cc/presentations/2013-09-11-cpp-seasoning/cpp-seasoning.pdf) or [the actual talk](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiTzoCxtd7zAhVJ2KQKHQ_DDRkQwqsBegQIERAB&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DqH6sSOr-yk8&usg=AOvVaw05PpTXVawOQVPckH4ygSyS). :smile: 
- Is `isBoundary[i]` **not** true for all j? I do not know/understand the indexing?! I would expect `i` to be the index of the degree of freedom and `j` refer to the coordinates or so.

ajaust reviewed (changes requested):
Thank you again for the adapter and sorry for the late response. I like your work, but nevertheless left some critical remarks. You do not have to fix everything now, but I think there are some things we have to think about and maybe create issues for them to track them. I can also help with some things if needed. I left some remarks in the code.

In short, these are some of the general remarks:

- Make function parameters constant where possible. [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#con1-by-default-make-objects-immutable)
- Mark functions that do not change any members as const. [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconst-fct)
- Trivial types do not have to be passed by reference. [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const)
- Could we add some simple test (maybe with a solver dummy)? Ideally with some test that might be run via a GitHub action (I would help with that). We can also move this into an issue for a future version.
- The module name `dune-precice`is good. I would call the repository `dune-precice` instead of `dune-adapter`. This would follow the standard nomenclature used by DUNE projects. It would make sense to change the layout as well by moving everything from the directory `dune-precice` in the root of the repository. This would be follow structure of other DUNE modules.
- `dune-precice` should not depend on `dune-elastodynamics` in my opinion as the adapter itself does not depend on it. Only the supplied example depends on `dune-elastodynamics` and `dune-precice`This means we should do something with the supplied example in `src/`.
- I did not try to compile or work with the supplied example yet.
- Document the dependency on a preCICE version. Does it work with preCICE v1 or v2? Does it work with both version?
- We should maybe add more documentation and some test (solverdummies)?!

davidscn commented:
@maxfirmbach do you have any idea which parts you want to address or if you want to address parts at all? Do you need help with some comments? If you don't want to keep working on this we could also merge and add a TODO list.

maxfirmbach commented:
I can address the changes to the code, but I work quite unregulary on the dune-adapter ... so it always takes some time. Considering the dependencies and the structure of the tutorial examples etc., I guess that's a part, which shouldn't be decided by me. If a merge + ToDo List is ok, I'm fine with that, I will work on the adapter from time to time.

davidscn commented:
Alright that's totally fine, then let's keep it as it is. I just want to prevent that things become stale.

maxfirmbach committed:
Restructure adapter and tutorials

maxfirmbach commented:
I restructured the adapter and the tutorial cases. Now there are two DUNE modules inside the `dune-adapter`, namely `dune-precice` containing the adapter (depending only on the core modules) and `dune-precice-howto` containing the tutorial case (and it's dependencies). For now I think that is a good structure and separation.

maxfirmbach committed:
First change requestions

maxfirmbach committed:
Second change requestions

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+      double old_time_;
+      int old_iter_;
+     
+      int mpi_rank_;  
```
```suggestion
      const int mpi_rank_;  
```
We only set the member in the constructor so we can make it `const`.

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+      int iteration_count = 0;
+      for(int i=0; i<dune_data.N(); i++) {
+        for(int j=0; j<dim; j++) {
+          if(is_coupling_boundary[i][j]) { dune_data[i][j] = precice_data[iteration_count]; ++iteration_count; }
```
```suggestion
          if(is_coupling_boundary[i][j]) { 
            dune_data[i][j] = precice_data[iteration_count]; 
            ++iteration_count; 
          }
```

Is this common/good practice in DUNE to have several statements in one line?

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+      int iteration_count = 0;
+      for(int i=0; i<dune_data.N(); i++) {
+        for(int j=0; j<dim; j++) {
+          if(is_coupling_boundary[i][j]) { precice_data[iteration_count] = dune_data[i][j]; ++iteration_count; }
```
```suggestion
          if(is_coupling_boundary[i][j]) { 
            precice_data[iteration_count] = dune_data[i][j]; 
            ++iteration_count; 
          }
```
Is this common/good practice in DUNE to have several statements in one line?

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+      int write_data_id_;
+      
+      int n_interface_nodes_;
+      std::vector<double> interface_nodes_positions_;
```
I think this should defined locally in the constructor instead of being a member variable because this vector is only used in the constructor, but not afterwards.
```suggestion
```

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+    auto coordinate = [] (auto x) { return x; };
+ 
+    interpolate(polynomial_basis, coordinates_of_boundary, coordinate, is_coupling_boundary_);
+
```
```suggestion
    std::vector<double> interface_nodes_positions;
```

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+
+    for(int i=0; i<coordinates_of_boundary.N(); i++) {
+      for(int j=0; j<dim; j++) {
+        if(is_coupling_boundary_[i][j]) interface_nodes_positions_.push_back(coordinates_of_boundary[i][j]);
```
```suggestion
        if(is_coupling_boundary_[i][j]) interface_nodes_positions.push_back(coordinates_of_boundary[i][j]);
```

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+
+    write_data_.resize(dim * n_interface_nodes_);
+    read_data_.resize(dim * n_interface_nodes_);
+    interface_nodes_positions_.resize(dim * n_interface_nodes_);
```
```suggestion
    interface_nodes_positions.resize(dim * n_interface_nodes_);
```

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+    std::cout << "preCICE:  Write ID on rank "    << mpi_rank_ << " = " << write_data_id_     << std::endl;  
+    std::cout << "preCICE:  Vertex size on rank " << mpi_rank_ << " = " << n_interface_nodes_ << std::endl;
+       
+    precice.setMeshVertices(mesh_id_, n_interface_nodes_, interface_nodes_positions_.data(), interface_nodes_ids_.data());
```
```suggestion
    precice.setMeshVertices(mesh_id_, n_interface_nodes_, interface_nodes_positions.data(), interface_nodes_ids_.data());
```

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+  double CouplingInterface<dim, VectorType, ParameterClass>::initialize(
+    const PolynomialBasis &polynomial_basis) 
+  {
+	mesh_id_       = precice.getMeshID(mesh_name_);
+	read_data_id_  = precice.getDataID(read_data_name_, mesh_id_);
+	write_data_id_ = precice.getDataID(write_data_name_, mesh_id_);
```
```suggestion
      mesh_id_       = precice.getMeshID(mesh_name_);
      read_data_id_  = precice.getDataID(read_data_name_, mesh_id_);
      write_data_id_ = precice.getDataID(write_data_name_, mesh_id_);
```
The indent is off. I hope I counted correctly in the suggestion (6 spaces).

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+  }
+      
+  template <int dim, typename VectorType, typename ParameterClass>
+  double CouplingInterface<dim, VectorType, ParameterClass>::advance(double computed_timestep_length)
```
```suggestion
  double CouplingInterface<dim, VectorType, ParameterClass>::advance(const double computed_timestep_length)
```

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+  
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  void CouplingInterface<dim, VectorType, ParameterClass>::read_blockvector_data(VectorType &dune_data)
```
```suggestion
  void CouplingInterface<dim, VectorType, ParameterClass>::read_blockvector_data(VectorType &dune_data) const
```

It should be possible to make this a constant function.

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+      template <typename PolynomialBasis>
+      double initialize(const PolynomialBasis &polynomial_basis);
+       
+      void read_blockvector_data(VectorType &dune_data);
```
```suggestion
      void read_blockvector_data(VectorType &dune_data) const;
```
Can't we make this constant?

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+      double initialize(const PolynomialBasis &polynomial_basis);
+       
+      void read_blockvector_data(VectorType &dune_data);
+      void send_blockvector_data(const VectorType &dune_data);
```
```suggestion
      void write_blockvector_data(const VectorType &dune_data);
```
Why is it called `send_blockvector_data`? It technically writes the data into the buffer. Data is only sent when `advance` is called. I think it would be possible to call the `write` functions of preCICE several times and even overwrite data in the buffer before data is actually send.

maxfirmbach replied:
Agreed, write is a way better naming for it!

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+
+#include <dune/functions/functionspacebases/interpolate.hh>
+
+#include <precice/SolverInterface.hpp>
```
Does it make sense to include it inside the `Dune::preCICE` namespace to keep the scope of preCICE more narrow?

maxfirmbach replied:
I'm not sure, does it make a big difference?

ajaust replied:
I am not sure what would be the best here. Including it within the namespace would hide preCICE and its `precice` namespace a bit more from users. From my perspective this would be something good since we want people to use the adapter. However, I did not find any examples regarding good and bad practices. This is for sure a small thing we can also easily change later. I just wanted to mention it.

Maybe @fsimonis has an idea/suggestion regarding the best practices regarding includes inside/outside namespaces?!

davidscn replied:
I think we can just keep this way, it is also more clear from an outside view (at least in my opinion), because you have all headers before starting to do anything.

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+  
+  
+  template <int dim, typename VectorType, typename ParameterClass>
+  void CouplingInterface<dim, VectorType, ParameterClass>::send_blockvector_data(const VectorType &dune_data)
```
The name needs to be changed if we decide exchange `send` by `write`.

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+      std::vector<double> read_data_;
+      std::vector<double> write_data_;
+      
+      std::vector<VectorType> old_state_data_;
+      double old_time_;
+      int old_iter_;
```
```suggestion
      struct {
        std::vector<VectorType> data;
        double time;
        int iter;
      } previous_solver_state_;
```
One could put this things in a struct to keep them together as they are tightly connected with each other.

maxfirmbach replied:
Yes, that's a good idea!

ajaust commented on dune-precice/dune-precice/couplinginterface.hh:
```diff
@@ -0,0 +1,264 @@
+    double &time,
+    int &iter)
+  {
+    for (int i=0; i<state_quantaties.size(); i++) {
```
```suggestion
    assert( state_quantaties.size() == old_state_data_.size() );
    for (int i=0; i<state_quantaties.size(); i++) {
```
Should we add an assertion here or is it guaranteed that the size of `state_quantaties` has not changed?

maxfirmbach replied:
I guess it won't hurt having one.

ajaust reviewed (commented):
Thank you very much for the work. I went through the main adapter part and did not check the dune-precice. I like the state of the adapter a lot. I have just some minor remarks.

maxfirmbach committed:
Third change requestions

davidscn commented:
I didn't follow the discussions, but I think this PR is ready to merge, @maxfirmbach any concerns? Is the tutorial ready as well?

ajaust commented:
I agree that it is ready. If there are problems, we can still discuss and change/fix things later.

maxfirmbach commented:
From my side there are no concerns, at least for now. The tutorial should also be ready.

davidscn commented:
Thank you for your contribution!
